consensus/driver/driver_test.go

```diff
@@ -0,0 +1,223 @@
+package driver_test
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/NethermindEth/juno/consensus/driver"
+	"github.com/NethermindEth/juno/consensus/mocks"
+	"github.com/NethermindEth/juno/consensus/tendermint"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/mock/gomock"
+)
+
+type expectedBroadcast struct {
+	proposals  []tendermint.Proposal[value, felt.Felt, felt.Felt]
+	prevotes   []tendermint.Prevote[felt.Felt, felt.Felt]
+	precommits []tendermint.Precommit[felt.Felt, felt.Felt]
+}
+
+type value uint64
+
+func (v value) Hash() felt.Felt {
+	return *new(felt.Felt).SetUint64(uint64(v))
+}
+
+type mockListener[M tendermint.Message[value, felt.Felt, felt.Felt]] struct {
+	ch chan M
+}
+
+func newMockListener[M tendermint.Message[value, felt.Felt, felt.Felt]](ch chan M) *mockListener[M] {
+	return &mockListener[M]{
+		ch: ch,
+	}
+}
+
+func (m *mockListener[M]) Listen() <-chan M {
+	return m.ch
+}
+
+type mockBroadcaster[M tendermint.Message[value, felt.Felt, felt.Felt]] struct {
+	wg                  sync.WaitGroup
+	mu                  sync.Mutex
+	broadcastedMessages []M
+}
+
+func (m *mockBroadcaster[M]) Broadcast(msg M) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.broadcastedMessages = append(m.broadcastedMessages, msg)
+	m.wg.Done()
+}
+
+func (m *mockBroadcaster[M]) SendMsg(addr felt.Felt, msg M) {
```

Comment:
Empty implementation intentional?

Also, receive address by reference, although since it is a test is not that important. The only thing that worries me in this scenario is making the code inconsistent by at copies being by reference and in other instances as by value.

Reply:
I'm wondering if we can remove this `SendMsg`, because the algorithm only needs to broadcast message, while direct communication between peers is handled inside the libp2p library.

Reply:
It makes sense to me